sled-agent/src/bootstrap/rack_secret.rs

```diff
@@ -0,0 +1,171 @@
+use std::convert::AsRef;
+use std::fmt::Debug;
+
+#[cfg(test)]
+use std::fmt;
+
+use p256::elliptic_curve::group::ff::PrimeField;
+use p256::{NonZeroScalar, ProjectivePoint, Scalar, SecretKey};
+use rand::rngs::OsRng;
+use serde::{Deserialize, Serialize};
+use vsss_rs::{Feldman, FeldmanVerifier, Share};
+
+/// A `RackSecret` is a shared secret used to perform a "rack-level" unlock.
+///
+/// Each server sled on an oxide rack contains up to 10 SSDs containing customer
+/// data and oxide specific data. We want to ensure that if a small number of
+/// disks or sleds get stolen, the data on them remains inaccessible to the
+/// thief.  We also want to ensure that successfully rebooting a sled or an
+/// entire rack does not require administrator intervention such as entering a
+/// password.
+///
+/// To provide the above guarantees we must ensure that all disks are encrypted,
+/// and that there is an automatic mechanism to retrieve the decryption key.
+/// Furthermore, we must guarantee that the the key retrieval mechanism is not
+/// available when a only a subset of sleds or disks from a rack are available
+/// to an attacker. The mechanism we use to provide these guarantees is based on
+/// <https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing>. A threshold secret
+/// is generated of which individual disk encryption keys are derived, and each
+/// of `N` server sleds receives one share of the secret. `K` of these shares
+/// must be combinded in order to reconstruct the shared secret such that disks
+/// may be decrypted. If fewer than `K` shares are available, no information
+/// about the secret may be recovered, and the disks cannot be decrypted. We
+/// call the threshold secret the `rack secret`.
+///
+/// Inside a rack then, the sleds cooperate over secure channels in order to
+/// retrieve key shares and reconstruct the `rack secret`, the resulting derived
+/// encryption keys, and unlock their own local storage. We call this procedure
+/// `rack unlock`. The establishment of secure channels and the ability to trust
+/// the validity of a participating peer is outside the scope of this particular
+/// type and orthogonal to its implementation.
+pub struct RackSecret {
+    secret: NonZeroScalar,
+}
+
+// This is a secret. Let's not print it outside of tests.
+#[cfg(test)]
+impl Debug for RackSecret {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.secret.as_ref().fmt(f)
+    }
+}
```

Comment:
Since this is part of the same crate and we're already making it conditionally compiled, it seems cleaner to just move this into the `mod test` block to me.

See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e4b30e9188c99ea63d6c719ba8159c56

This keeps the test-only code more isolated from the non-test code.